woocommerce/woocommerce-gutenberg-products-block
# Allow extensions to extend cart item price

Fixes #3605.

Related PR in WC Subscriptions: 3971-gh-woocommerce/woocommerce-subscriptions

This PR does a couple of things:
* Introduces a new filter (available through the `cartItemPrice` key) that allows extensions to modify the way cart item prices are displayed in the Cart and Checkout blocks. Notice this filter passes a string and expects a string, so extensions can add text before or after the price, but they can't replace it with another component. There is also a check to verify the filtered string contains the price, so extensions can't remove it by mistake.
* In f51199b854a03cfc35d40b3d5a8f16c75642f5d7, it changes the value used to display the cart item price. Until now, we were using the formula `product_price * quantity`. But we already get a `totals` property from the API with the value calculated so we can use that one. This has a couple of consequences:
  * Subscriptions sets the cart item total to the sign-up fee. That was lost in blocks because we were ignoring the total and instead were calculating it in the client side. With this PR it should be fixed.
  _(Sidenote: the client side calculation of cart item totals was also problematic in PAO because it was ignoring the price of the add-ons)_
  * :warning: Given that we are no longer making the item price calculation client side, optimistic updates no longer work. Debounce is still there, only change is that optimistic updates will not be visible.

I'm adding the :warning: emoji because this is a feature loss. I think optimistic updates need to be re-thought from the base. The way it worked until now was showing the user a value that was different from the one in the server. IMO this is very dangerous and has proven problematic in Subscriptions and PAO so far.

### Screenshots

![imatge](https://user-images.githubusercontent.com/3616980/106001761-a5f8f900-60b0-11eb-893e-577ebcbb9488.png)

![imatge](https://user-images.githubusercontent.com/3616980/106001812-b3ae7e80-60b0-11eb-9acc-4de305e19b7e.png)

### How to test the changes in this Pull Request:

1. Checkout out this branch in WC Subscriptions: `fix/cart-item-price-filter`.
2. Add a Subscription product to your cart with a sign-up fee.
3. Go to the Cart and Checkout blocks and verify the price for that cart item is shown as `Due today $1.23` (in Cart) and `$1.23 due today` (in Checkout).
4. Test with Subscriptions without sign-up fee, with regular products, etc. to verify all cases work fine.
